base_projects/Drivetrain/src/main/cpp/Robot.cpp

```diff
@@ -0,0 +1,367 @@
+  // TODO use std::exit or EndCompetition once CTRE bug is fixed
+  std::set_terminate([]() { std::_Exit(0); });
+#endif
+}
+
+void Robot::AddMotorController(
+    int port, std::string controller, bool inverted,
+    std::vector<std::unique_ptr<frc::SpeedController>>* controllers) {
+  if (controller == "TalonSRX" || controller == "VictorSPX" ||
+      controller == "TalonFX") {
+    if (controller == "TalonSRX") {
+      controllers->emplace_back(std::make_unique<WPI_TalonSRX>(port));
+    } else if (controller == "TalonFX") {
+      controllers->emplace_back(std::make_unique<WPI_TalonFX>(port));
+    } else {
+      controllers->emplace_back(std::make_unique<WPI_VictorSPX>(port));
+    }
+    dynamic_cast<WPI_BaseMotorController*>(controllers->back().get())
+        ->SetInverted(inverted);
+  } else if (controller == "SPARK MAX (Brushless)" ||
+             controller == "SPARK MAX (Brushed)") {
+    if (controller == "SPARK MAX (Brushless)") {
+      controllers->emplace_back(std::make_unique<rev::CANSparkMax>(
+          port, rev::CANSparkMax::MotorType::kBrushless));
+    } else {
+      controllers->emplace_back(std::make_unique<rev::CANSparkMax>(
+          port, rev::CANSparkMax::MotorType::kBrushed));
+    }
+    static_cast<rev::CANSparkMax*>(controllers->back().get())
+        ->SetInverted(inverted);
+  } else {
+    controllers->emplace_back(std::make_unique<frc::Spark>(port));
+    static_cast<frc::Spark*>(controllers->back().get())->SetInverted(inverted);
+  }
+}
+
+void Robot::SetupEncoders(const std::string& controllerName,
+                          frc::SpeedController* controller,
+                          bool encoderInverted,
+                          const std::vector<int>& encoderPorts,
+                          std::unique_ptr<CANCoder>& cancoder,
+                          std::unique_ptr<rev::CANEncoder>& canencoder,
+                          std::unique_ptr<frc::Encoder>& encoder,
+                          std::function<double()>& position,
+                          std::function<double()>& rate) {
+  if (m_encoderType == "Built-In") {
+    if (controllerName == "TalonSRX" || controllerName == "TalonFX") {
+      if (controllerName == "TalonSRX") {
+        dynamic_cast<WPI_BaseMotorController*>(controller)
+            ->ConfigSelectedFeedbackSensor(FeedbackDevice::QuadEncoder);
+      } else {
+        dynamic_cast<WPI_BaseMotorController*>(controller)
+            ->ConfigSelectedFeedbackSensor(FeedbackDevice::IntegratedSensor);
+      }
+      dynamic_cast<WPI_BaseMotorController*>(controller)
+          ->SetSensorPhase(encoderInverted);
+      position = [&, this] {
+        return dynamic_cast<WPI_BaseMotorController*>(controller)
+                   ->GetSelectedSensorPosition() /
+               m_cpr;
+      };
+      rate = [&, this] {
+        return dynamic_cast<WPI_BaseMotorController*>(controller)
+                   ->GetSelectedSensorVelocity() /
+               m_cpr / .1;
```

Comment:
This magic number should be a constant or documented somehow.
```suggestion
               m_cpr / 0.1;
```

Reply:
Would it be better to use the units library to specify this is converting the rawSensorUnits/100ms into rotations/second?